kafka_checked/src/main.rs

```diff
@@ -0,0 +1,236 @@
+struct Obstime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    tbtime: Vec<Tbtime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <tbtime>...</tbtime>
+struct Tbtime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    sensor: Vec<Sensor>,
+    kvtextdata: Option<Vec<Kvtextdata>>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <sensor>...</sensor>
+struct Sensor {
+    #[serde(rename = "@val")]
+    val: i32,
+    level: Vec<Level>,
+}
+/// Represents <level>...</level>
+#[derive(Debug, Deserialize)]
+struct Level {
+    #[serde(rename = "@val")]
+    val: i32,
+    kvdata: Option<Vec<Kvdata>>,
+}
+/// Represents <kvdata>...</kvdata>
+#[derive(Debug, Deserialize)]
+struct Kvdata {
+    #[serde(rename = "@paramid")]
+    paramid: i32,
+    original: Option<f64>,
+    corrected: Option<f64>,
+    controlinfo: Option<String>,
+    useinfo: Option<String>,
+    cfailed: Option<String>,
+}
+/// Represents <kvtextdata>...</kvtextdata>
+#[derive(Debug, Deserialize)]
+struct Kvtextdata {
+    _paramid: Option<i32>,
+    _original: Option<String>,
+}
+#[derive(Debug, Deserialize)]
+struct Tsid {
+    station: i32,
+    paramid: i32,
+    typeid: i32,
+    sensor: i32,
+    level: i32,
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+
+    assert!(args.len() >= 5, "not enough args passed in, at least group name, host, user, dbname needed, optionally password");
+
+    let mut connect_string = format!("host={} user={} dbname={}", &args[2], &args[3], &args[4]);
+    if args.len() > 5 {
+        connect_string.push_str(" password=");
+        connect_string.push_str(&args[5]);
+    }
+    println!("Connecting to database with string: {:?} \n", connect_string);
+
+    // Connect to the database.
+    let (client, connection) = tokio_postgres::connect(connect_string.as_str(), NoTls).await?;
+
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("connection error: {e}");
+        }
+    });
+
+    let group_string = format!("{}", &args[1]);
+    println!("Creating kafka consumer with group name: {:?} \n", group_string);
+    // NOTE: reading from the 4 redundant kafka queues, but only reading the checked data (other topics exists)
+    let mut consumer = Consumer::from_hosts(vec![
+        "kafka2-a1.met.no:9092".to_owned(),
+        "kafka2-a2.met.no:9092".to_owned(),
+        "kafka2-b1.met.no:9092".to_owned(),
+        "kafka2-b2.met.no:9092".to_owned(),
+    ])
+    .with_topic_partitions("kvalobs.production.checked".to_owned(), &[0, 1])
+    .with_fallback_offset(FetchOffset::Earliest)
+    .with_group(group_string)
+    .with_offset_storage(Some(GroupOffsetStorage::Kafka))
+    .create()
+    .unwrap();
+
+    // Consume the kafka queue infinitely 
+    loop {
+        for ms in consumer.poll().unwrap().iter() {
```

Comment:
This is from the example in the documentation: https://docs.rs/kafka/latest/kafka/consumer/index.html 
I had pondered if it should be more explicit in its handling of the connection and messages, from my undestanding it might then make sense to use a slightly lower level way of talking to the queue: https://docs.rs/kafka/latest/kafka/client/index.html 
I will look more into this, and the changes needed / ramifications.

Reply:
Obligatory "I'm not familiar with this library", but I think using the higher level API is fine, just handle the Result properly. They wouldn't have made the function return `Result` if you weren't supposed to handle it, they just didn't bother to set up error handling in their example code

Reply:
Is this unwrap actually ok? https://docs.rs/kafka/latest/src/kafka/consumer/mod.rs.html#155
While it returns a result, it doesn't actually contain an error. Unsure if it can be empty message sets...

Reply:
What do you mean that doesn't contain an error? That code you linked definitely contains an error.

Perhaps you don't think `UnknownTopicOrPartition` is likely to happen, but then you should at least expect. Since the inner code generates that error from a `None`, I personally would suspect that it can happen when anything goes wrong communicating with the kafka queue.

Reply:
I think its just me just being a complete rust newb and thinking that this result from Poll didn't explicitly contain an error. I read superficially and was thinking that .Poll would explicicly say Restult<messagesets, error> (my misunderstanding being that its either a result or an error: https://doc.rust-lang.org/std/result/). But I see that the underlying functions are perfectly capable of returning an error, so I should have read more deeply. Definetly thinking that connection issues can be expected when talking to the kafka queue... concerned about retries / reconnects, and how that should be handled.

Reply:
That result type is one they've defined in the crate so they don't have to specify their error type every time, it's a common practice in Rust crates. This is the definition:

```rust
pub type Result<T> = std::result::Result<T, Error>;
```

It's actually impossible to create a std::result::Result without specifying an Error type.